### PR TITLE
Desktop: Add "retry all" option to synchronisation page

### LINKED ISSUE #2688: Add option to "Retry all" when multiple items failed decrypting
When an item cannot be decrypted, it appears on the synchronisation page, and there's a "Retry" link next to it to tell the decryption worker to process the item again.

If there's more than 1 item, there should be a "Retry all" button to send all the items for decryption in one go. That button should be at the top of the list.

### PR DESCRIPTION
**Pull request by fhfuih**
Try to fix #2688 
 **Note: the codes in this PR is not tested yet. I think I need some help on how to trigger an error in the synchronisation page, so I can see the program outcome.**
Also have to update localization template

### CODE REVIEW

**Comment by tessus** [comment]
If you want to link an issue, please use:

```
fixes #2688 
closes #2688 
```

**Comment by fhfuih** [comment]
> If you want to link an issue, please use either one of these:
> 
> ```
> fixes #2688 
> closes #2688 
> try to fix #2688
> ```

Oh I see... Thanks!

**Comment by tessus** [comment]
Yep, otherwise github doesn't recognize the "link", if there's something between `closes|fixes|close|fix` and the number `#xxxx`.

You can read more here: https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue

**Comment by fhfuih** [comment]
@tessus I am just curious whether you or any other collaborators will do the code review and testing stuff for good first issues? Since no on seems interested in answering my question of how I can test the functionality myself on the issue page and discourse😂 It's a small enhancement with just a few lines of code though, but I feel like being as much responsible as I can to my PR.

**Comment by tessus** [comment]
@fhfuih Code reviews are usually done by Laurent, mic704b, CalebJohn, or myself (or any other person who has a good grasp of the code base).
But testing should be done by you. There's a testing framework and test cases should be written for code that isn't UI stuff.
In case of UI code, the coder should provide a checklist what tests were done manually.

There are so many topics on discourse lately that it's highly possible that your question was missed.

**Comment by fhfuih** [comment]
OK thanks!

File: ElectronClient/gui/StatusScreen.jsx [hunk 1]
```diff
@@ -114,7 +136,14 @@ class StatusScreenComponent extends React.Component {
 				sectionsHtml.push(renderSectionHtml(i, section));
 			}
 
-			return <div>{sectionsHtml}</div>;
+			let retryAllHtml = renderRetryAllHtml(); // must be called after renderSectionHtml
```
Comment (laurent22) [inline ElectronClient/gui/StatusScreen.jsx]: Please refactor the code so that this hack is not needed.

Comment (fhfuih) [reply ElectronClient/gui/StatusScreen.jsx]: Sure. BTW is there any coding style regarding `forEach` and `for(i < array.length)`? I notice a mixture of two in the project.

Comment (fhfuih) [reply ElectronClient/gui/StatusScreen.jsx]: I just refactored my code using `forEach` then, guess it's not a big deal 😂

**Comment by laurent22** [comment]
I will post the same comment as in the other PR:

* * *

The feature request was for a Retry All button when items cannot be decrypted. But with this code, you'll add a Retry All button that will retry everything, including decryption, downloading resources again, etc.

Also in general it's best not to put business logic like that in the view. All this should be in services/report.js. Maybe have a look at this file and see how the sections are built, then add your Retry All buttons (there should be two of them) at the right place.

The advantage of doing it that way is that it will work in the mobile app too.

**Comment by fhfuih** [comment]
> I will post the same comment as in the other PR:
> 
> The feature request was for a Retry All button when items cannot be decrypted. But with this code, you'll add a Retry All button that will retry everything, including decryption, downloading resources again, etc.
> 
> Also in general it's best not to put business logic like that in the view. All this should be in services/report.js. Maybe have a look at this file and see how the sections are built, then add your Retry All buttons (there should be two of them) at the right place.
> 
> The advantage of doing it that way is that it will work in the mobile app too.

Yeah that indeed makes sense. I will try to solve the problem in another way. Thank you!

**Comment by tessus** [comment]
@fhfuih what's the status?

**Comment by fhfuih** [comment]
@tessus Hi I have updated my branct with new commits shown above. I didn't reply then because I thought you will get notifications when new commits are present in the PR. Sorry I about that! 

The new commit consists of backend logic and buttons for both desktop and mobile interface. Again I didn't see the outcome since I don't know how to trigger a decryption error yet.

**Comment by tessus** [comment]
Yes, I saw the commits, but I can't know, if you are done. So something like `Addressed comments/issues` or something similar goes a long way to clear things up.

**Comment by fhfuih** [comment]
@tessus
I see... Sorry that it's my first time trying to resolve an issue of another project. 

fixes #2688

**Comment by laurent22** [comment]
Did you test your changes? Maybe it's me but this looks plain broken:

image

The Retry All button is floating there with random margin on the left, and no clear relation with the items to be retried below.

Another issue is that the "Retry All" button would retry all items **simultaneously**, so if someone has 1000 items it would try to decrypt 1000 items at the same time.

I'm going to fix this for you as we need this feature and don't want to spend more time on this.

**Comment by fhfuih** [comment]
Like what I have said earlier, I would love to have tested if you have told me how to trigger a decryption error. Or if you need to get this feature done you can indeed fix it yourself anyway.

**Comment by laurent22** [comment]
What did you try though? Logically to create an encryption error you need to corrupt the encrypted data, i.e. add or remove a character from it. We could have provided info if you had told us what you tried, what didn't work, etc. But explaining everything from beginning to end should not be necessary.

**Comment by fhfuih** [comment]
Well I have tried to read the source code and tried to figure out what this "decryption/encryption" means and where I can enable "decryption/encryption" and where I can find the encrypted file. But it's a hard and time-consuming process to understand the logic given that I need to care about GSoC proposal on other ideas at the same time. Of course I knew that eventually I would have to corrupt a encryption file as this is the most straightforward part of the entire procedures thank you for your explanation. 

If you have looked closer you should have notice that I have asked the same question in this PR **twice**

> Note: the codes in this PR is not tested yet. I think I need some help on how to trigger an error in the synchronisation page, so I can see the program outcome.
> The new commit consists of backend logic and buttons for both desktop and mobile interface. Again I didn't see the outcome since I don't know how to trigger a decryption error yet.

And the same question in the issue thread
> However, I don't know how to trigger a decryption failure. I just tried disconnected from the internet but it just says Completed: <time here> with no fetch items x/x

And the same question in the [Joplin forum thread](https://discourse.joplinapp.org/t/introducing-fhfuih/6736)

Our conversation could have been smoother even if you have throw me this reply earlier when you have seen my question at the first place because you may have not understood me.

Nobody wants to be a nuisance and asking the same dumb question over and over again. I already feel very tired and disrespected of receiving only comments on my PR but not a single word regarding this question for days over and over again. 

So please take it and implement yourself. You can always communicate with yourself without any efforts and waiting. Hope you have gained some valuable pride when you tried hard to ignore my words to let me know you are picky about questions to answer.

I'm done with it.

**Comment by laurent22** [comment]
The thing is, it's a time-consuming process for everybody. You need to look after your pull request and proposal, and we need to review 50+ pull requests and many proposals.

This is why we expect GSoC candidates to do their own work and look out for solutions by themselves. If you get stuck, then it's when we can help, but we can't give you the solution on a plate, or else indeed it would be simpler to do it ourselves.

Our mistake this year, which will fix next year, is that we should make it clear the pull requests should be complete and working, and with test units. We don't have the resources to accept WIP pull requests, or provide step by step instructions for testing or development.

**Comment by fhfuih** [comment]
The thing is, I totally understood that you guys are busy and might not reply everyone promptly. The thing I didn't understand and found it irritating is when you guys finally had time to check this out and reply me, you only commented on my codes but simply ignored that question, literally ignoring the question. And that situation happened over and over again. Even if you have merely said "we expect GSoC candidates to do their own work and look out for solutions by themselves" earlier instead of being that ignorant, it would be fine to me. There's absolutely nothing to do about you guys being busy in my complaint above.

> If you get stuck, then it's when we can help, but we can't give you the solution on a plate,

Plus, I don't know how do you define "get stuck". I definitely wanted to code the new feature myself. I just wanted to know which button to click to toggle encryption and where to find the files. And I thought I got stuck. And I didn't ask you for a solution to coding the feature.

**Pull request closed without merging**